src/main/scala/org/ergoplatform/nodeView/mempool/ErgoMemPoolReader.scala

```diff
@@ -1,57 +1,28 @@
 package org.ergoplatform.nodeView.mempool
 
 import org.ergoplatform.modifiers.mempool.ErgoTransaction
-import org.ergoplatform.nodeView.mempool.ErgoMemPool._
 import scorex.core.transaction.MempoolReader
 import scorex.util.ModifierId
 
-import scala.collection.concurrent.TrieMap
-import scala.concurrent.{Future, Promise}
-import scala.util.Success
+import scala.collection.immutable.TreeMap
 
 trait ErgoMemPoolReader extends MempoolReader[ErgoTransaction] {
 
-  val unconfirmed: TrieMap[ModifierId, ErgoTransaction]
+  val unconfirmed: TreeMap[(ModifierId, Long), ErgoTransaction]
 
-  /**
-    * Map stores current state of waiting for query building
-    * value - promise of result and set of all transactions of request
-    * key - set of transactions that are waiting for the assembly
-    */
-  protected[mempool] var waitedForAssembly: Map[Set[ModifierId], (Promise[MemPoolResponse], Seq[ModifierId])] = Map.empty
-
-  override def modifierById(modifierId: ModifierId): Option[ErgoTransaction] = unconfirmed.get(modifierId)
-
-  override def contains(id: ModifierId): Boolean = unconfirmed.contains(id)
+  override def contains(id: ModifierId): Boolean = unconfirmed.values.exists(_.id == id)
 
   override def getAll(ids: Seq[ModifierId]): Seq[ErgoTransaction] = ids.flatMap(modifierById)
 
   override def size: Int = unconfirmed.size
 
-  def take(limit: Int): Iterable[ErgoTransaction] =
-    unconfirmed.values.toSeq.take(limit)
-
-  //TODO rework option.get
-  protected def completeAssembly(txs: Iterable[ErgoTransaction]): Unit = synchronized {
-    val txsIds = txs.map(_.id)
-    val newMap = waitedForAssembly.flatMap(p => {
-      val ids = p._1
-      val newKey = ids -- txsIds
-      // filtering fully-built queries and completing of a promise
-      if (newKey.isEmpty) {
-        val (promise, allIds) = p._2
-        promise complete Success(allIds.map(id => modifierById(id).get))
-        None
-      } else {
-        Some(newKey -> p._2)
-      }
-    })
-    waitedForAssembly = newMap
-  }
-
-  def waitForAll(ids: MemPoolRequest): Future[MemPoolResponse] = synchronized {
-    val promise = Promise[Seq[ErgoTransaction]]
-    waitedForAssembly = waitedForAssembly.updated(ids.toSet, (promise, ids))
-    promise.future
-  }
+  def getAll: Seq[ErgoTransaction] = unconfirmed.values.toSeq
+
+  /** Returns all transactions resided in pool sorted by priority */
+  def getAllPrioritized: Seq[ErgoTransaction] = unconfirmed.values.toList.reverse
+
+  def take(limit: Int): Iterable[ErgoTransaction] = unconfirmed.values.toSeq.take(limit)
+
+  def modifierById(modifierId: ModifierId): Option[ErgoTransaction] = unconfirmed.values.find(_.id == modifierId)
```

Comment:
How often this method is called, doesn't look scalable.

Reply:
@ergomorphic, this method is not used at all, as well as contains()

Reply:
It is used in NodeViewSynchronizer in Scorex, when a node process inv or receive modifier. It looks quite important to have a scalable implementation here

Reply:
@catena2w, Ok, I've improved its performance.